From ea62b989b3ccb4bdf99cd42cfc17f08723c63407 Mon Sep 17 00:00:00 2001
From: Eric Garver <eric@garver.life>
Date: Wed, 29 Oct 2025 16:33:17 -0400
Subject: [PATCH 47/48] v2.4.0: chore(ipXtables): simplify icmp match fragments

We can use ICMP_TYPES/ICMPV6_TYPES to get the codes and make rule
generation generic while also supporting more types that don't have name
support in iptables, e.g. mld.

(cherry picked from commit 1562687ae6ca3cbcf6ece25126af9116adfb897b)
---
 src/firewall/core/ipXtables.py | 61 +++++++++++++++-------------------
 1 file changed, 27 insertions(+), 34 deletions(-)

diff --git a/src/firewall/core/ipXtables.py b/src/firewall/core/ipXtables.py
index 339840c37ba4..efe292b3f638 100644
--- a/src/firewall/core/ipXtables.py
+++ b/src/firewall/core/ipXtables.py
@@ -27,10 +27,12 @@ from firewall.core.logger import log
 from firewall.functions import tempFile, readfile, splitArgs, check_mac, portStr, \
                                check_single_address, check_address, normalizeIP6
 from firewall import config
-from firewall.errors import FirewallError, INVALID_PASSTHROUGH, INVALID_RULE, UNKNOWN_ERROR, INVALID_ADDR
+from firewall.errors import FirewallError, INVALID_PASSTHROUGH, INVALID_RULE, UNKNOWN_ERROR, INVALID_ADDR, \
+                            INVALID_ICMPTYPE
 from firewall.core.rich import Rich_Accept, Rich_Reject, Rich_Drop, Rich_Mark, Rich_NFLog, \
                                Rich_Masquerade, Rich_ForwardPort, Rich_IcmpBlock, Rich_Tcp_Mss_Clamp
 from firewall.core.base import DEFAULT_ZONE_TARGET
+from firewall.core.icmp import ICMP_TYPES, ICMPV6_TYPES
 import string
 
 POLICY_CHAIN_PREFIX = ""
@@ -598,37 +600,14 @@ class ip4tables(object):
         return rules
 
     def supported_icmp_types(self, ipv=None):
-        """Return ICMP types that are supported by the iptables/ip6tables command and kernel"""
-        ret = [ ]
-        output = ""
-        try:
-            output = self.__run(["-p",
-                                 "icmp" if self.ipv == "ipv4" else "ipv6-icmp",
-                                 "--help"])
-        except ValueError as ex:
-            if self.ipv == "ipv4":
-                log.debug1("iptables error: %s" % ex)
-            else:
-                log.debug1("ip6tables error: %s" % ex)
-        lines = output.splitlines()
-
-        in_types = False
-        for line in lines:
-            #print(line)
-            if in_types:
-                line = line.strip().lower()
-                splits = line.split()
-                for split in splits:
-                    if split.startswith("(") and split.endswith(")"):
-                        x = split[1:-1]
-                    else:
-                        x = split
-                    if x not in ret:
-                        ret.append(x)
-            if self.ipv == "ipv4" and line.startswith("Valid ICMP Types:") or \
-               self.ipv == "ipv6" and line.startswith("Valid ICMPv6 Types:"):
-                in_types = True
-        return ret
+        supported = set()
+
+        if ipv is None or self.ipv == "ipv4":
+            supported.update(ICMP_TYPES.keys())
+        if ipv is None or self.ipv == "ipv6":
+            supported.update(ICMPV6_TYPES.keys())
+
+        return list(supported)
 
     def build_default_tables(self):
         # nothing to do, they always exist
@@ -1350,6 +1329,20 @@ class ip4tables(object):
 
         return rules
 
+    def _icmp_types_fragment(self, icmp_type):
+        if self.ipv == "ipv4" and icmp_type in ICMP_TYPES:
+            _type, _code, _omit_code = ICMP_TYPES[icmp_type]
+            _type_str = str(_type) if _omit_code else str(_type) + "/" + str(_code)
+            return ["-m", "icmp", "--icmp-type", _type_str]
+        elif self.ipv == "ipv6" and icmp_type in ICMPV6_TYPES:
+            _type, _code, _omit_code = ICMPV6_TYPES[icmp_type]
+            _type_str = str(_type) if _omit_code else str(_type) + "/" + str(_code)
+            return ["-m", "icmp6", "--icmpv6-type", _type_str]
+        else:
+            raise FirewallError(
+                INVALID_ICMPTYPE, f"ICMP type {icmp_type} not supported by {self.name}"
+            )
+
     def build_policy_icmp_block_rules(self, enable, policy, ict, rich_rule=None):
         table = "filter"
         _policy = self._fw.policy.policy_base_chain_name(policy, table, POLICY_CHAIN_PREFIX)
@@ -1357,10 +1350,10 @@ class ip4tables(object):
 
         if self.ipv == "ipv4":
             proto = [ "-p", "icmp" ]
-            match = [ "-m", "icmp", "--icmp-type", ict.name ]
+            match = self._icmp_types_fragment(ict.name)
         else:
             proto = [ "-p", "ipv6-icmp" ]
-            match = [ "-m", "icmp6", "--icmpv6-type", ict.name ]
+            match = self._icmp_types_fragment(ict.name)
 
         rules = []
         if self._fw.policy.query_icmp_block_inversion(policy):
-- 
2.47.3

